runtime/bindings/python/numpy_interop.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "./numpy_interop.h"
+
+#include "./binding.h"
+
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include "numpy/arrayobject.h"
+
+namespace iree::python::numpy {
+
+namespace {
+
+int internal_import_array() {
+  import_array1(-1);
+  return 0;
+}
+
+}  // namespace
+
+void InitializeNumPyInterop() {
+  if (internal_import_array() < 0) {
+    throw py::import_error("numpy.core.multiarray failed to import");
+  }
+}
+
+int ConvertHalElementTypeToNumPyTypeNum(iree_hal_element_type_t t) {
```

Comment:
The function this replaces (`MapElementTypeToDType`) had a comment written after quite a bit of debugging (https://github.com/openxla/iree/pull/12880). Worth extracting something from that comment to this new function?

```
  // See:
  //   * https://numpy.org/doc/stable/reference/arrays.dtypes.html
  //   * https://docs.python.org/3/c-api/arg.html#numbers
  //
  // Single letter codes can be ambiguous across platforms, so prefer explicit
  // bit depth values, ("Type strings: Any string in numpy.sctypeDict.keys()").
  // See https://github.com/pybind/pybind11/issues/1908
```

Reply:
Yes, but that was just because it had to use the (very) old string format stuff. This is just a strict mapping of enumeration values and has no ambiguity that wouldn't be obvious given the overall C API usage here.